### ISSUE TITLE
Gemma3 and Video Understanding

### ISSUE THREAD

**Issue by AmeenAli** [modelscope/ms-swift#6199]
is Video Understanding finetuning for Gemma3 supported?

it seems that we have a hard assert:

https://github.com/modelscope/ms-swift/blob/main/swift/llm/template/template/gemma.py#L102

accepting only images (not as with qwen2 VL)

i can submit a PR if you agree on this, otherwise i would love to hear more clarifications on how to finetune gemma3 on video.

**Comment by AmeenAli** [comment]
I Simply changes the Gemma3VisionTemplate to:

`
class Gemma3VisionTemplate(Gemma3Template):
    boi_token_id = 255999
    placeholder_tokens = ['<start_of_image>']

    def replace_tag(self, media_type: Literal['image', 'video', 'audio'], index: int,
                    inputs: StdTemplateInputs) -> List[Context]:

        from qwen_vl_utils import fetch_image, fetch_video
        assert media_type in {'image', 'video'}
        # kwargs = {'image_patch_size': 14} #if self.version == 'v3' else {}

        if media_type == 'image':
            return ['<start_of_image>']
        if media_type == 'video':
            video = inputs.videos[index]

            video, video_kwargs = fetch_video({'video': video}, return_video_sample_fps=True)
            fps = video.shape[0]
            
            if isinstance(video, torch.Tensor):
                video = video.to(torch.uint8)
            inputs.videos[index] = video
            tokens = [self.tokenizer.boi_token] * fps
            return tokens

    def _encode(self, inputs: StdTemplateInputs) -> Dict[str, Any]:
        from transformers.models.gemma3.processing_gemma3 import Gemma3ProcessorKwargs

        encoded = super()._encode(inputs)
        if inputs.images:
            input_ids = encoded['input_ids']
            labels = encoded['labels']
            loss_scale = encoded.get('loss_scale', None)
            idx_list = findall(input_ids, self.boi_token_id)
            img_tokens = self._tokenize(self.processor.full_image_sequence)
            input_ids, labels = self._extend_tokens(input_ids, labels, idx_list, lambda _: img_tokens)
            loss_scale = self._extend_loss_scale(loss_scale, idx_list, lambda _: img_tokens)

            # TODO: customize
            processor_kwargs = Gemma3ProcessorKwargs._defaults['images_kwargs']
            image_inputs = self.processor.image_processor(inputs.images, **processor_kwargs)
            image_inputs['pixel_values'] = torch.as_tensor(np.array(image_inputs['pixel_values']))
            image_inputs.pop('num_crops')

            array_ids = np.array(input_ids)
            mm_token_type_ids = np.zeros_like(input_ids)
            mm_token_type_ids[array_ids == self.processor.image_token_id] = 1
            encoded['token_type_ids'] = mm_token_type_ids.tolist()
            encoded['input_ids'] = input_ids
            encoded['pixel_values'] = image_inputs['pixel_values']
            encoded['labels'] = labels
            encoded['loss_scale'] = loss_scale
        if inputs.videos:
            input_ids = encoded['input_ids']
            labels = encoded['labels']
            loss_scale = encoded.get('loss_scale', None)
            idx_list = findall(input_ids, self.tokenizer.boi_token_id)
            
            processor_kwargs = Gemma3ProcessorKwargs._defaults['images_kwargs']
            
            fps = inputs.videos[0].shape[0]

            img_tokens = self._tokenize(self.processor.full_image_sequence)
            input_ids, labels, loss_scale = self._extend_tokens(input_ids, labels, idx_list,
                                                            lambda _: img_tokens)
    
            all_imgs = []
            for i in range(fps):
                img = torch.as_tensor(np.array(self.processor.image_processor([inputs.videos[0][i]], **processor_kwargs)['pixel_values']))
                all_imgs.append(img.unsqueeze(0))
            pixel_values = torch.cat(all_imgs, dim=0).squeeze()

            array_ids = np.array(input_ids)

            mm_token_type_ids = np.zeros_like(input_ids)
            mm_token_type_ids[array_ids == self.processor.image_token_id] = 1
            encoded['token_type_ids'] = mm_token_type_ids.tolist()
            encoded['input_ids'] = input_ids
            encoded['pixel_values'] = pixel_values
            encoded['labels'] = labels
            encoded['loss_scale'] = loss_scale
        return encoded
`

Not sure if this solves the issue, I am investigating it

**Comment by AmeenAli** [comment]
the above fix does not seem to be working
@Jintao-Huang any thoughts on this?

**Comment by slin000111** [comment]
> the above fix does not seem to be working [@Jintao-Huang](https://github.com/Jintao-Huang) any thoughts on this?

It seems this model's chat template doesn't support video.
The following code is copied from the chat_template.json file.
```text
... {%- if item['type'] == 'image' -%}\n                {{ '<start_of_image>' }}\n            {%- elif item['type'] == 'text' -%}\n                {{ item['text'] | trim }}\n            {%- endif -%}\n ...
```

**Comment by AmeenAli** [comment]
it does supports video
simply by appending multiple image tokens

https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma3/processing_gemma3.py#L102

@slin000111
